File: qa/support.py
import argparse
import json
from urllib.parse import urlsplit

from playwright.sync_api import sync_playwright


PAGES = [
    "index.html",
    "crypto-summary.html",
    "crypto-summary-ar.html",
    "bitcoin-from-zero.html",
    "bitcoin-from-zero-ar.html",
    "proof-of-stake.html",
    "solana.html",
    "sui.html",
    "monero-under-the-hood.html",
    "monero-from-zero-ar.html",
    "zcash.html",
    "zero-knowledge-from-zero.html",
    "getting-monero.html",
]
ADDRESSES = {
    "XMR": "49BWStdJJVEfPRDQTmCtrT39QSqXNWLUGNvTzU7cMfsMHTG1ozR1YFQPCKb65oduji5isbVAowiwVFPpseHv9Dg3875rB9G",
    "BTC": "bc1q7avt57ekp002lzq2untsr5lpqy9hsfqudf7a5h",
    "LTC": "ltc1qrvecg74m4elgm9j2el78zgau5a4l3x245jwqmj",
    "ZEC": "u1hpydpv0n36jvucwevy6477auq5ml5wjgzfrwq20xsau78mj9x576fkx7cd2q23f35lr0wctznsj27tnxh0anujlcdn45sqsvqm0kshl09982f4tmpa64npzezu5m3d7wm4vv6uqjlvkdda9lzavn8yqfgmzaykgcxt5lqmme5qr5epwt",
    "DASH": "XjZo388Ct5tssNwC9BDbzXqC52YXqyUPsz",
    "ETH": "0x56102e5dc2bCE0ab5766e87a2363cea93FbB7D4d",
    "SOL": "8jUUGAtcTW7ZxGMUPtuJGsmyep5oxmcrBENFZ8TDoiBm",
}
NETWORKS = {
    "XMR": "Monero network",
    "BTC": "Bitcoin network",
    "LTC": "Litecoin network",
    "ZEC": "Zcash Unified Address",
    "DASH": "Dash network",
    "ETH": "Ethereum network",
    "SOL": "Solana network",
}
REFERRALS = {
    "Bybit": "https://www.bybit.com/invite?ref=MWBL9%230&medium=referral&utm_campaign=evergreen",
    "Binance": "https://www.binance.com/referral/earn-together/refer2earn-usdc/claim?hl=en&ref=GRO_28502_WQILP&utm_source=referral_entrance",
}
DISCLOSURE = "These are referral links. I may receive a benefit if you use them."


def check(condition, message):
    if not condition:
        raise AssertionError(message)


def origin_for(base_url):
    parsed = urlsplit(base_url)
    return f"{parsed.scheme}://{parsed.netloc}"


def verify_includes(page, base_url):
    results = []
    for name in PAGES:
        response = page.goto(f"{base_url}/{name}", wait_until="networkidle")
        check(response and response.status == 200, f"{name}: HTTP load failed")
        page.wait_for_function("window.__support?.dialog", timeout=5000)
        check(page.locator(".support-trigger").count() == 1, f"{name}: expected one Donate button")
        check(page.locator("dialog.support-dialog").count() == 1, f"{name}: expected one support dialog")
        results.append(name)
    return results


def open_dialog(page):
    trigger = page.locator(".support-trigger")
    trigger.focus()
    trigger.click()
    dialog = page.locator("dialog.support-dialog")
    check(dialog.get_attribute("open") is not None, "Dialog did not open")
    page.wait_for_function("document.activeElement?.id === 'support-tab-xmr'", timeout=1000)
    return trigger, dialog


def verify_currencies(page, dialog):
    copied = []
    for code, expected in ADDRESSES.items():
        tab = dialog.locator(f"[data-currency='{code}']")
        check(tab.count() == 1, f"{code}: currency option missing")
        tab.click()
        check(dialog.locator(".support-address").inner_text() == expected, f"{code}: displayed address changed")
        check(dialog.locator(".support-identity-copy span").inner_text() == NETWORKS[code], f"{code}: wrong network label")
        copy_button = dialog.locator(".support-copy")
        check(copy_button.is_enabled(), f"{code}: copy unexpectedly disabled")
        copy_button.click()
        page.wait_for_function(
            "code => document.querySelector('.support-status')?.textContent.includes(code)",
            arg=code,
            timeout=2000,
        )
        clipboard = page.evaluate("navigator.clipboard.readText()")
        check(clipboard == expected, f"{code}: clipboard value changed")
        copied.append(code)
    check(copied == ["XMR", "BTC", "LTC", "ZEC", "DASH", "ETH", "SOL"], f"Unexpected copied currencies {copied}")
    return copied


def verify_referrals(dialog):
    links = dialog.locator(".support-referral-links a")
    check(links.count() == 2, "Expected two referral links")
    for index, (name, expected) in enumerate(REFERRALS.items()):
        link = links.nth(index)
        check(link.inner_text().startswith(name), f"{name}: label missing")
        check(link.get_attribute("href") == expected, f"{name}: URL changed")
        rel = set((link.get_attribute("rel") or "").split())
        check(rel == {"sponsored", "noopener", "noreferrer"}, f"{name}: rel attributes wrong")
    check(dialog.locator(".support-disclosure").inner_text() == DISCLOSURE, "Referral disclosure changed")
    hrefs = dialog.locator("a").evaluate_all("links => links.map(link => link.getAttribute('href'))")
    check(hrefs == list(REFERRALS.values()), "Unexpected dialog link or transaction link present")
    remote_images = dialog.locator("img").evaluate_all(
        "images => images.map(image => image.getAttribute('src')).filter(src => /^https?:/i.test(src || ''))"
    )
    check(not remote_images, f"Remote image dependency present {remote_images}")


def verify_keyboard(page, trigger, dialog):
    page.keyboard.press("Escape")
    check(dialog.get_attribute("open") is None, "Escape did not close dialog")
    check(trigger.evaluate("el => el === document.activeElement"), "Focus did not return to Donate button")
    trigger.click()
    dialog.locator(".support-close").click()
    check(dialog.get_attribute("open") is None, "Close button did not close dialog")
    check(trigger.evaluate("el => el === document.activeElement"), "Close button did not restore focus")


def verify_copy_error(browser, base_url):
    context = browser.new_context(viewport={"width": 1000, "height": 800})
    page = context.new_page()
    page.goto(f"{base_url}/index.html", wait_until="networkidle")
    page.wait_for_function("window.__support?.dialog", timeout=5000)
    page.evaluate(
        """() => {
            Object.defineProperty(navigator, 'clipboard', {
                configurable: true,
                value: {writeText: () => Promise.reject(new Error('blocked'))}
            });
            Document.prototype.execCommand = () => false;
        }"""
    )
    page.locator(".support-trigger").click()
    page.locator(".support-copy").click()
    page.wait_for_function("document.querySelector('.support-status')?.textContent.startsWith('Copy failed')", timeout=2000)
    check("manually" in page.locator(".support-status").inner_text(), "Copy error did not provide manual fallback")
    context.close()


def verify_mobile(browser, base_url):
    results = []
    for width in (390, 320):
        context = browser.new_context(viewport={"width": width, "height": 844})
        page = context.new_page()
        page.goto(f"{base_url}/index.html", wait_until="networkidle")
        page.wait_for_function("window.__support?.dialog", timeout=5000)
        check(page.locator(".support-trigger").is_visible(), f"Donate button hidden at {width}px")
        page.locator(".support-trigger").click()
        metrics = page.locator("dialog.support-dialog").evaluate(
            """dialog => {
                const box = dialog.getBoundingClientRect();
                const tabs = [...dialog.querySelectorAll('.support-tab')].map(tab => {
                    const rect = tab.getBoundingClientRect();
                    return {left: rect.left, right: rect.right, top: rect.top, bottom: rect.bottom};
                });
                const tablist = dialog.querySelector('.support-tabs');
                return {
                    left: box.left,
                    right: box.right,
                    top: box.top,
                    bottom: box.bottom,
                    viewport: innerWidth,
                    tabs,
                    tabOverflow: tablist.scrollWidth - tablist.clientWidth,
                    addressSize: parseFloat(getComputedStyle(dialog.querySelector('.support-address')).fontSize)
                };
            }"""
        )
        check(metrics["left"] >= 0 and metrics["right"] <= metrics["viewport"], f"Mobile dialog overflow at {width}px: {metrics}")
        check(len(metrics["tabs"]) == 7, f"Expected seven mobile tabs at {width}px")
        check(
            all(
                tab["left"] >= metrics["left"]
                and tab["right"] <= metrics["right"]
                and tab["top"] >= metrics["top"]
                and tab["bottom"] <= metrics["bottom"]
                for tab in metrics["tabs"]
            ),
            f"Currency tab escaped visible dialog at {width}px: {metrics}",
        )
        check(metrics["tabOverflow"] <= 1, f"Currency grid overflow at {width}px: {metrics}")
        check(metrics["addressSize"] == 12, f"Address text is not 12px at {width}px")
        check(page.evaluate("document.body.scrollWidth <= innerWidth + 1"), f"Mobile page overflow at {width}px")
        results.append({"width": width, "tabs": len(metrics["tabs"]), "address_px": metrics["addressSize"]})
        context.close()
    return results


def verify_dark(browser, base_url):
    context = browser.new_context(viewport={"width": 1100, "height": 800})
    page = context.new_page()
    page.goto(f"{base_url}/index.html", wait_until="networkidle")
    page.wait_for_function("window.__support?.dialog", timeout=5000)
    page.evaluate("document.documentElement.dataset.theme = 'dark'")
    page.locator(".support-trigger").click()
    colors = page.locator("dialog.support-dialog").evaluate(
        "dialog => ({dialog: getComputedStyle(dialog).backgroundColor, page: getComputedStyle(document.documentElement).getPropertyValue('--page').trim()})"
    )
    check(colors["dialog"] not in {"rgb(255, 255, 255)", "#fff", "white"}, f"Dark dialog remained white {colors}")
    check(page.locator(".support-trigger").is_visible(), "Donate button disappeared in dark theme")
    context.close()


def verify_rtl(browser, base_url):
    context = browser.new_context(viewport={"width": 1100, "height": 800})
    context.grant_permissions(["clipboard-read", "clipboard-write"], origin=origin_for(base_url))
    page = context.new_page()
    page.goto(f"{base_url}/crypto-summary-ar.html", wait_until="networkidle")
    page.wait_for_function("window.__support?.dialog", timeout=5000)
    trigger = page.locator(".support-trigger")
    check(trigger.inner_text() == "تبرع", "Arabic Donate label missing")
    trigger.click()
    dialog = page.locator("dialog.support-dialog")
    check(dialog.evaluate("el => getComputedStyle(el).direction") == "rtl", "Dialog did not inherit RTL")
    check(dialog.locator(".support-address").evaluate("el => getComputedStyle(el).direction") == "ltr", "Address did not remain LTR")
    dialog.locator("[data-currency='BTC']").click()
    dialog.locator(".support-copy").click()
    page.wait_for_function("document.querySelector('.support-status')?.textContent.includes('BTC')", timeout=2000)
    check(page.evaluate("navigator.clipboard.readText()") == ADDRESSES["BTC"], "RTL copy changed BTC address")
    page.keyboard.press("Escape")
    check(trigger.evaluate("el => el === document.activeElement"), "RTL Escape did not restore focus")
    context.close()


def run(base_url):
    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(channel="chrome")
        context = browser.new_context(viewport={"width": 1280, "height": 900})
        context.grant_permissions(["clipboard-read", "clipboard-write"], origin=origin_for(base_url))
        page = context.new_page()
        included = verify_includes(page, base_url)
        page.goto(f"{base_url}/index.html", wait_until="networkidle")
        trigger, dialog = open_dialog(page)
        copied = verify_currencies(page, dialog)
        verify_referrals(dialog)
        verify_keyboard(page, trigger, dialog)
        context.close()
        verify_copy_error(browser, base_url)
        mobile = verify_mobile(browser, base_url)
        verify_dark(browser, base_url)
        verify_rtl(browser, base_url)
        browser.close()
    return {
        "included_pages": included,
        "copied_currencies": copied,
        "ethereum": "checksum valid and copied exactly",
        "referrals": "pass",
        "keyboard_and_focus": "pass",
        "copy_error_fallback": "pass",
        "mobile": mobile,
        "dark": "pass",
        "rtl": "pass",
    }


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--base-url", default="http://127.0.0.1:8934")
    args = parser.parse_args()
    print(json.dumps(run(args.base_url.rstrip("/")), indent=2, ensure_ascii=False))

File: assets/support.js
(function () {
  'use strict';

  if (window.__support) return;

  var root = document.documentElement;
  var rtl = root.dir === 'rtl' || root.lang === 'ar';
  var labels = rtl ? {
    trigger: 'تبرع',
    title: 'ادعم Crypto in Brief',
    intro: 'ساعد في إبقاء الشروحات والتجارب متاحة للجميع.',
    choose: 'اختر العملة',
    address: 'عنوان الاستلام',
    copy: 'نسخ العنوان',
    copied: 'تم نسخ عنوان',
    failed: 'تعذر النسخ. حدد العنوان وانسخه يدويا.',
    close: 'إغلاق',
    referrals: 'روابط الإحالة',
    open: 'فتح'
  } : {
    trigger: 'Donate',
    title: 'Support Crypto in Brief',
    intro: 'Help keep the explainers and working labs available to everyone.',
    choose: 'Choose a currency',
    address: 'Receiving address',
    copy: 'Copy address',
    copied: 'Copied',
    failed: 'Copy failed. Select the address and copy it manually.',
    close: 'Close',
    referrals: 'Referral links',
    open: 'Open'
  };

  var currencies = [
    {
      code: 'XMR',
      name: 'Monero',
      network: 'Monero network',
      address: '49BWStdJJVEfPRDQTmCtrT39QSqXNWLUGNvTzU7cMfsMHTG1ozR1YFQPCKb65oduji5isbVAowiwVFPpseHv9Dg3875rB9G',
      image: 'assets/coins/monero.svg'
    },
    {
      code: 'BTC',
      name: 'Bitcoin',
      network: 'Bitcoin network',
      address: 'bc1q7avt57ekp002lzq2untsr5lpqy9hsfqudf7a5h',
      image: 'assets/coins/bitcoin.svg'
    },
    {
      code: 'LTC',
      name: 'Litecoin',
      network: 'Litecoin network',
      address: 'ltc1qrvecg74m4elgm9j2el78zgau5a4l3x245jwqmj'
    },
    {
      code: 'ZEC',
      name: 'Zcash',
      network: 'Zcash Unified Address',
      address: 'u1hpydpv0n36jvucwevy6477auq5ml5wjgzfrwq20xsau78mj9x576fkx7cd2q23f35lr0wctznsj27tnxh0anujlcdn45sqsvqm0kshl09982f4tmpa64npzezu5m3d7wm4vv6uqjlvkdda9lzavn8yqfgmzaykgcxt5lqmme5qr5epwt',
      image: 'assets/coins/zcash.svg'
    },
    {
      code: 'DASH',
      name: 'Dash',
      network: 'Dash network',
      address: 'XjZo388Ct5tssNwC9BDbzXqC52YXqyUPsz'
    },
    {
      code: 'ETH',
      name: 'Ethereum',
      network: 'Ethereum network',
      address: '0x56102e5dc2bCE0ab5766e87a2363cea93FbB7D4d',
      image: 'assets/coins/ethereum.svg'
    },
    {
      code: 'SOL',
      name: 'Solana',
      network: 'Solana network',
      address: '8jUUGAtcTW7ZxGMUPtuJGsmyep5oxmcrBENFZ8TDoiBm',
      image: 'assets/coins/solana.svg'
    }
  ];

  var referrals = [
    ['Bybit', 'https://www.bybit.com/invite?ref=MWBL9%230&medium=referral&utm_campaign=evergreen'],
    ['Binance', 'https://www.binance.com/referral/earn-together/refer2earn-usdc/claim?hl=en&ref=GRO_28502_WQILP&utm_source=referral_entrance']
  ];

  function make(tag, className, text) {
    var element = document.createElement(tag);
    if (className) element.className = className;
    if (text != null) element.textContent = text;
    return element;
  }

  function addMark(target, currency, compact) {
    if (currency.image) {
      var image = make('img', 'support-coin-image');
      image.src = currency.image;
      image.alt = '';
      image.width = compact ? 22 : 34;
      image.height = compact ? 22 : 34;
      target.appendChild(image);
    } else {
      target.appendChild(make('span', 'support-coin-type', currency.code));
    }
  }

  function copyText(text) {
    if (navigator.clipboard && window.isSecureContext) {
      return navigator.clipboard.writeText(text).then(function () {
        return true;
      }).catch(function () {
        return false;
      });
    }
    return Promise.resolve(false);
  }

  function selectAddress(element) {
    var selection = window.getSelection();
    if (!selection) return;
    var range = document.createRange();
    range.selectNodeContents(element);
    selection.removeAllRanges();
    selection.addRange(range);
  }

  var target = document.querySelector('.site-nav') || document.querySelector('.site-header');
  if (!target) return;

  var trigger = make('button', 'support-trigger', labels.trigger);
  trigger.type = 'button';
  trigger.setAttribute('aria-haspopup', 'dialog');

  if (target.classList.contains('site-nav')) {
    var themeButton = target.querySelector('.theme-button');
    target.insertBefore(trigger, themeButton || null);
  } else {
    var switcher = target.querySelector('.switch');
    target.insertBefore(trigger, switcher || null);
  }

  var dialog = make('dialog', 'support-dialog');
  dialog.setAttribute('aria-labelledby', 'support-title');
  var sheet = make('div', 'support-sheet');
  var heading = make('header', 'support-heading');
  var headingCopy = make('div');
  var kicker = make('span', 'support-kicker', 'Crypto in Brief');
  var title = make('h2', '', labels.title);
  title.id = 'support-title';
  var intro = make('p', '', labels.intro);
  headingCopy.appendChild(kicker);
  headingCopy.appendChild(title);
  headingCopy.appendChild(intro);
  var closeButton = make('button', 'support-close', labels.close);
  closeButton.type = 'button';
  heading.appendChild(headingCopy);
  heading.appendChild(closeButton);

  var body = make('div', 'support-body');
  var tabs = make('div', 'support-tabs');
  tabs.setAttribute('role', 'tablist');
  tabs.setAttribute('aria-label', labels.choose);
  var panel = make('section', 'support-panel');
  panel.id = 'support-panel';
  panel.setAttribute('role', 'tabpanel');
  panel.tabIndex = 0;

  var identity = make('div', 'support-identity');
  var identityMark = make('span', 'support-identity-mark');
  var identityCopy = make('span', 'support-identity-copy');
  var coinName = make('strong');
  var network = make('span');
  identityCopy.appendChild(coinName);
  identityCopy.appendChild(network);
  identity.appendChild(identityMark);
  identity.appendChild(identityCopy);

  var addressLabel = make('span', 'support-address-label', labels.address);
  var address = make('code', 'support-address');
  address.dir = 'ltr';
  var copyButton = make('button', 'support-copy', labels.copy);
  copyButton.type = 'button';
  var status = make('p', 'support-status');
  status.setAttribute('role', 'status');
  status.setAttribute('aria-live', 'polite');
  var addressBlock = make('div', 'support-address-block');
  addressBlock.appendChild(addressLabel);
  addressBlock.appendChild(address);
  addressBlock.appendChild(copyButton);
  addressBlock.appendChild(status);
  panel.appendChild(identity);
  panel.appendChild(addressBlock);
  body.appendChild(tabs);
  body.appendChild(panel);

  var referralSection = make('section', 'support-referrals');
  referralSection.appendChild(make('h3', '', labels.referrals));
  var referralLinks = make('div', 'support-referral-links');
  referrals.forEach(function (referral) {
    var link = make('a', '', referral[0] + ' ' + labels.open);
    link.href = referral[1];
    link.target = '_blank';
    link.rel = 'sponsored noopener noreferrer';
    referralLinks.appendChild(link);
  });
  referralSection.appendChild(referralLinks);
  referralSection.appendChild(make('p', 'support-disclosure', 'These are referral links. I may receive a benefit if you use them.'));

  sheet.appendChild(heading);
  sheet.appendChild(body);
  sheet.appendChild(referralSection);
  dialog.appendChild(sheet);
  document.body.appendChild(dialog);

  var buttons = [];
  var selected = 0;
  var returnFocus = null;

  function selectCurrency(index, focus) {
    selected = index;
    var currency = currencies[index];
    buttons.forEach(function (button, buttonIndex) {
      var current = buttonIndex === index;
      button.setAttribute('aria-selected', current ? 'true' : 'false');
      button.tabIndex = current ? 0 : -1;
    });
    identityMark.replaceChildren();
    addMark(identityMark, currency, false);
    coinName.textContent = currency.name + ' · ' + currency.code;
    network.textContent = currency.network;
    address.textContent = currency.address;
    address.setAttribute('aria-label', currency.code + ' ' + labels.address);
    copyButton.disabled = false;
    copyButton.textContent = labels.copy;
    status.textContent = '';
    panel.setAttribute('aria-labelledby', buttons[index].id);
    if (focus) buttons[index].focus();
  }

  currencies.forEach(function (currency, index) {
    var button = make('button', 'support-tab');
    button.type = 'button';
    button.id = 'support-tab-' + currency.code.toLowerCase();
    button.setAttribute('role', 'tab');
    button.setAttribute('aria-controls', panel.id);
    button.setAttribute('data-currency', currency.code);
    var mark = make('span', 'support-tab-mark');
    addMark(mark, currency, true);
    var copy = make('span', 'support-tab-copy');
    copy.appendChild(make('strong', '', currency.code));
    copy.appendChild(make('small', '', currency.name));
    button.appendChild(mark);
    button.appendChild(copy);
    button.addEventListener('click', function () {
      selectCurrency(index, false);
    });
    button.addEventListener('keydown', function (event) {
      var next = null;
      if (event.key === 'ArrowDown' || event.key === 'ArrowRight') next = (index + 1) % buttons.length;
      if (event.key === 'ArrowUp' || event.key === 'ArrowLeft') next = (index - 1 + buttons.length) % buttons.length;
      if (event.key === 'Home') next = 0;
      if (event.key === 'End') next = buttons.length - 1;
      if (next != null) {
        event.preventDefault();
        selectCurrency(next, true);
      }
    });
    buttons.push(button);
    tabs.appendChild(button);
  });

  copyButton.addEventListener('click', function () {
    var currency = currencies[selected];
    copyText(currency.address).then(function (copied) {
      if (copied) {
        status.textContent = labels.copied + ' ' + currency.code;
        return;
      }
      status.textContent = labels.failed;
      selectAddress(address);
    });
  });

  function openDialog() {
    returnFocus = document.activeElement;
    if (typeof dialog.showModal === 'function') dialog.showModal();
    else dialog.setAttribute('open', '');
    requestAnimationFrame(function () {
      buttons[selected].focus();
    });
  }

  function closeDialog() {
    if (typeof dialog.close === 'function') dialog.close();
    else {
      dialog.removeAttribute('open');
      if (returnFocus && returnFocus.focus) returnFocus.focus();
    }
  }

  trigger.addEventListener('click', openDialog);
  closeButton.addEventListener('click', closeDialog);
  dialog.addEventListener('close', function () {
    if (returnFocus && returnFocus.focus) returnFocus.focus();
  });
  dialog.addEventListener('click', function (event) {
    if (event.target === dialog) closeDialog();
  });

  selectCurrency(0, false);
  window.__support = {
    dialog: dialog,
    currencies: currencies.slice(),
    referrals: referrals.slice(),
    open: openDialog,
    close: closeDialog
  };
})();
